### PR TITLE
Remove all instances of low-level DH use in libssl

### PR DESCRIPTION
**Pull request by mattcaswell**
To prepare for the forthcoming deprecation of all the remaining low-level DH functions (see #13138), we remove all such usage from libssl. 

There are a few instances where a DH object is passed as part of the public API. In those cases we deprecate the functions/macros where that is the case and provider EVP_PKEY based replacements.

### CODE REVIEW

File: doc/man3/SSL_CTX_set_tmp_dh_callback.pod [hunk 1]
```diff
@@ -22,93 +32,73 @@ SSL_set_tmp_dh_callback, SSL_set_tmp_dh
...
+If there is no certificate (e.g. for PSK ciphersuites), then it it will be
+consistent with the size of the negotiated symmetric cipher key.
+
+Applications may supply their own DH parameters instead of using the built-in
+values. To do so they should call SSL_CTX_set0_tmp_dh_pkey() or
+SSL_set0_tmp_dh_pkey() to supply the parameters for the B<SSL_CTX> or B<SSL>
+respectively. The parameters should be supplied in the I<dhpkey> argument as
+an B<EVP_PKEY> containg DH parameters. Ownership of the I<dhpkey> value is
+passed to the B<SSL_CTX> or B<SSL> object as a result of this call, and so the
+caller should not free it if the function call is succesful.
```
Comment (kroeckx) [inline doc/man3/SSL_CTX_set_tmp_dh_callback.pod:76]: I think the text should discourage the use of those functions, and that they should use the built-in parameters instead.

File: test/sslapitest.c [hunk 2]
```diff
@@ -8045,7 +8046,193 @@ static int test_ssl_dup(void)
...
+ * Test  3: Explicit SSL_CTX auto on
+ * Test  4: Explicit SSL auto on
+ * Test  5: Explicit SSL_CTX auto off, custom DH params via EVP_PKEY
+ * Test  6: Explicit SSL auto off, custom DH params via EVP_PKEY
+ *
+ * The following are testing deprecated APIs, so we only run them if available
+ * Test  7: Explicit SSL_CTX auto off, custom DH params via DH
+ * Test  8: Explicit SSL auto off, custom DH params via DH
+ * Test  9: Explicit SSL_CTX auto off, custom DH params via callback
+ * Test 10: Explicit SSL auto off, custom DH params via callback
```
Comment (kroeckx) [inline test/sslapitest.c:8126]: I would like to see some tests that confirm that we select the correct DH parameters depending on the certificate/cipher.

File: apps/s_server.c [hunk 3]
```diff
@@ -3004,20 +3005,26 @@ static void print_connection_info(SSL *con)
...
     if ((bio = BIO_new_file(dhfile, "r")) == NULL)
         goto err;
-    ret = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
- err:
+
+    decoderctx
+        = OSSL_DECODER_CTX_new_by_EVP_PKEY(&dhpkey, "PEM", "DH",
+                                           NULL, NULL);
+    if (decoderctx != NULL)
+        OSSL_DECODER_from_bio(decoderctx, bio);
```
Comment (levitte) [inline apps/s_server.c:3021]: There is now a `load_keyparams` in libapps...  does it still make sense to have a specialized `load_dh_params`?

File: include/openssl/ssl.h.in [hunk 4]
```diff
@@ -2130,12 +2141,14 @@ void SSL_set_default_read_buffer_len(SSL *s, size_t len);
...
-                             DH *(*dh) (SSL *ssl, int is_export,
-                                        int keylength));
+DEPRECATEDIN_3_0(
+        void SSL_CTX_set_tmp_dh_callback(SSL_CTX *ctx,
+                                         DH *(*dh) (SSL *ssl, int is_export,
+                                                    int keylength)))
+DEPRECATEDIN_3_0(void SSL_set_tmp_dh_callback(SSL *ssl,
+                                              DH *(*dh) (SSL *ssl,
+                                                         int is_export,
+                                                         int keylength)))
```
Comment (levitte) [inline include/openssl/ssl.h.in:2151]:
``` C
#  ifndef OPENSSL_NO_DEPRECATED_3_0
OSSL_DEPRECATEDIN_3_0
void SSL_CTX_set_tmp_dh_callback(SSL_CTX *ctx,
                                 DH *(*dh) (SSL *ssl, int is_export,
                                            int keylength));
OSSL_DEPRECATEDIN_3_0
void SSL_set_tmp_dh_callback(SSL *ssl,
                             DH *(*dh) (SSL *ssl, int is_export,
                                        int keylength));
#  endif
```

File: ssl/statem/statem_clnt.c [hunk 5]
```diff
@@ -3079,44 +3067,48 @@ static int tls_construct_cke_dhe(SSL *s, WPACKET *pkt)
...
     if (ssl_derive(s, ckey, skey, 0) == 0) {
         /* SSLfatal() already called */
         goto err;
     }
 
     /* send off the data */
-    prime_len = BN_num_bytes(DH_get0_p(dh_clnt));
+
+    /* Generate encoding of server key */
+    encoded_pub_len = EVP_PKEY_get1_tls_encodedpoint(ckey, &encoded_pub);
```
Comment (levitte) [inline ssl/statem/statem_clnt.c:3078]: Travis complains about `EVP_PKEY_get1_tls_encodedpoint`...  missing an include file or has it been renamed by now?

https://travis-ci.com/github/openssl/openssl/jobs/405072122#L3847-L3850

File: doc/man3/SSL_CTX_set_tmp_dh_callback.pod (hunk 1 above)
Comment (mattcaswell) [reply doc/man3/SSL_CTX_set_tmp_dh_callback.pod:76]: I've strengthed the discouragement of this.

File: test/sslapitest.c (hunk 2 above)
Comment (mattcaswell) [reply test/sslapitest.c:8126]: I've added a new test in sslapitest.c for this.

File: apps/s_server.c (hunk 3 above)
Comment (mattcaswell) [reply apps/s_server.c:3021]: I've made `load_dh_params` a simple wrapper around `load_keyparams` which additionally confirms that we have DH parameters.

File: include/openssl/ssl.h.in (hunk 4 above)
Comment (mattcaswell) [reply include/openssl/ssl.h.in:2151]: Fixed.

File: ssl/statem/statem_clnt.c (hunk 5 above)
Comment (mattcaswell) [reply ssl/statem/statem_clnt.c:3078]: This was the result of a merge error when I rebased. Fixed now.

**Comment by mattcaswell** [comment]
All comments addressed. I've rebased and pushed the updates.

The new test reveals that if we use a ciphersuite without a cert and select a 128-bit symmetric cipher, then we end up using 1024-bit DH params. This seems very small.

**Comment by t8m** [comment]
> All comments addressed. I've rebased and pushed the updates.
> 
> The new test reveals that if we use a ciphersuite without a cert and select a 128-bit symmetric cipher, then we end up using 1024-bit DH params. This seems very small.

Yep, we should use at least 2048 bit DH params unless 1024 bit cert is used.

**Comment by mattcaswell** [comment]
Fix ups pushed addressing a mem leak reported by travis. I also made a tweak to the documentation.

**Comment by kroeckx** [comment]
> And what about 256 bit symmetric ciphers?

Note that it's not because it's using AES256 that it reaches the
256 bit security level.

I currently see:
openssl ciphers -s -v DEFAULT:@SECLEVEL=5
TLS_AES_256_GCM_SHA384         TLSv1.3 Kx=any      Au=any   Enc=AESGCM(256) Mac=AEAD
TLS_CHACHA20_POLY1305_SHA256   TLSv1.3 Kx=any      Au=any   Enc=CHACHA20/POLY1305(256) Mac=AEAD
ECDHE-ECDSA-AES256-GCM-SHA384  TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AESGCM(256) Mac=AEAD
ECDHE-RSA-AES256-GCM-SHA384    TLSv1.2 Kx=ECDH     Au=RSA   Enc=AESGCM(256) Mac=AEAD
DHE-RSA-AES256-GCM-SHA384      TLSv1.2 Kx=DH       Au=RSA   Enc=AESGCM(256) Mac=AEAD
ECDHE-ECDSA-CHACHA20-POLY1305  TLSv1.2 Kx=ECDH     Au=ECDSA Enc=CHACHA20/POLY1305(256) Mac=AEAD
ECDHE-RSA-CHACHA20-POLY1305    TLSv1.2 Kx=ECDH     Au=RSA   Enc=CHACHA20/POLY1305(256) Mac=AEAD
DHE-RSA-CHACHA20-POLY1305      TLSv1.2 Kx=DH       Au=RSA   Enc=CHACHA20/POLY1305(256) Mac=AEAD
ECDHE-ECDSA-AES256-SHA384      TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AES(256)  Mac=SHA384
ECDHE-RSA-AES256-SHA384        TLSv1.2 Kx=ECDH     Au=RSA   Enc=AES(256)  Mac=SHA384
DHE-RSA-AES256-SHA256          TLSv1.2 Kx=DH       Au=RSA   Enc=AES(256)  Mac=SHA256

And I'm not sure all of them actually reach the 256 bit security
level that level 5 is supposed to reach.

In theory, if you want to reach the 256 bit security level, it
should be a 15360 key, which is really slow. We should not use
such a large parameters with the auto setting. We probably don't
even have such a large named group. My recommendation would be
to go for the 8192 bit key, reaching the 192 bit security level. I
think trying to reach anything higher than the 192 bit security level
using DHE doesn't make any sense.

You can actually argue that if the security level that's set allows
a 1024 (level 1) or 2048 bit (level 2) key, that can use it. But I
think that if we can reach the 128 bit security level (level 3) that
we should do that.

**Comment by kroeckx** [comment]
> In theory, if you want to reach the 256 bit security level, it
> should be a 15360 key, which is really slow. We should not use
> such a large parameters with the auto setting. We probably don't
> even have such a large named group. My recommendation would be
> to go for the 8192 bit key, reaching the 192 bit security level. I
> think trying to reach anything higher than the 192 bit security level
> using DHE doesn't make any sense.

It seems we even don't support DH parameters larger then 10000.

**Comment by t8m** [comment]
Let's be realistic. With non-anonymous ciphersuites (regardless of whether AES-128 or AES-256 is used) we would select 2k DH parameters if 2k RSA cert is used which is still the most common case. I do not think we need to try to be more strict than that with anonymous ciphersuites.

**Comment by t8m** [comment]
And/or even better we could set the DH parameter size for anonymous ciphersuites based on the SECLEVEL setting with the low limit of 2048.

**Comment by mattcaswell** [comment]
I've rebased this to resolve a conflict with master. Also fixed a travis issue and added a commit removing deprecated DH usage from ssltest_old (also resolving travis issues).

File: doc/man3/SSL_CTX_set_tmp_dh_callback.pod [hunk 6]
```diff
@@ -22,93 +32,77 @@ SSL_set_tmp_dh_callback, SSL_set_tmp_dh
...
-supply at least 2048-bit parameters in the callback.
+should not generate the parameters on the fly. DH parameters can be reused, as
+the actual key is newly generated during the negotiation.
+
+Typically applications should use well know DH parameters that have built-in
+support in OpenSSL. The macros SSL_CTX_set_dh_auto() and SSL_set_dh_auto()
+configure OpenSSL to use the default built-in DH parameters for the B<SSL_CTX>
+and B<SSL> objects respectively. Passing a value of 1 in the I<onoff> parameter
+switches the feature on, and passing a value of 0 switches it off. The default
+setting is off.
```
Comment (t8m) [inline doc/man3/SSL_CTX_set_tmp_dh_callback.pod:63]: Hmm, should we make it on by default? I suppose that would be for another PR and OTC decision.

File: ssl/s3_lib.c [hunk 7]
```diff
@@ -3776,22 +3768,21 @@ long ssl3_callback_ctrl(SSL *s, int cmd, void (*fp) (void))
     int ret = 0;
 
     switch (cmd) {
-#ifndef OPENSSL_NO_DH
+#if !defined(OPENSSL_NO_DH ) && !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Comment (t8m) [inline ssl/s3_lib.c:3771]: Nit: extra space after OPENSSL_NO_DH

File: ssl/statem/statem_srvr.c [hunk 8]
```diff
@@ -2592,21 +2586,21 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
...
-            goto err;
-        }
-
         EVP_PKEY_free(pkdh);
         pkdh = NULL;
 
-        DH_get0_pqg(dh, &r[0], NULL, &r[1]);
-        DH_get0_key(dh, &r[2], NULL);
+        /* These BIGNUMs need to be freed when we're finished */
+        freer = 1;
```
Comment (t8m) [inline ssl/statem/statem_srvr.c:2593]: Instead of using this freer flag initialize the r array with NULLs. This is incorrect because theoretically any of the EVP_PKEY_get_bn_param could fail leaving the rest of the r uninitialized.

**Review by t8m — CHANGES_REQUESTED** [review]
Some minor fixes needed.

**Comment by t8m** [comment]
Also Travis failure is relevant.

File: apps/s_server.c [hunk 9]
```diff
@@ -3003,20 +3004,16 @@ static void print_connection_info(SSL *con)
...
+    EVP_PKEY *dhpkey = load_keyparams(dhfile, 0, "DH parameters");
 
-    if ((bio = BIO_new_file(dhfile, "r")) == NULL)
-        goto err;
-    ret = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
- err:
-    BIO_free(bio);
-    return ret;
+    if (dhpkey != NULL && !EVP_PKEY_is_a(dhpkey, "DH")) {
+        EVP_PKEY_free(dhpkey);
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply apps/s_server.c]: I've refactored this based on the commit from #13317 - so this PR is now dependent on that one.

File: doc/man3/SSL_CTX_set_tmp_dh_callback.pod (hunk 6 above)
Comment (mattcaswell) [reply doc/man3/SSL_CTX_set_tmp_dh_callback.pod:63]: Yes, possibly we should. But, as you say, as another PR and pending an OTC decision.

File: ssl/statem/statem_srvr.c (hunk 8 above)
Comment (mattcaswell) [reply ssl/statem/statem_srvr.c:2593]: Err, they are already initialised to NULL earlier on in this function. We still need to use the `freer` variables, because other codepaths populate the r array with BIGNUMs that should *no*t be freed. However, this one populates it with values that *should* be freed.

**Comment by mattcaswell** [comment]
> Also Travis failure is relevant.

Hopefully addressed. Lets see how it gets on now.

File: ssl/statem/statem_srvr.c (hunk 8 above)
Comment (t8m) [reply ssl/statem/statem_srvr.c:2593]: Ah, I overlooked the initialization in the unchanged code.

File: ssl/tls_depr.c [hunk 10]
```diff
@@ -143,5 +143,21 @@ HMAC_CTX *ssl_hmac_get0_HMAC_CTX(SSL_HMAC *ctx)
 {
     return ctx->old_ctx;
 }
+
+/* Some deprecated public APIs pass DH objects */
+# ifndef OPENSSL_NO_DH
+EVP_PKEY *ssl_dh_to_pkey(DH *dh)
+{
+    EVP_PKEY *ret;
+    if (dh == NULL)
```
Comment (slontis) [inline ssl/tls_depr.c:152]: blankline

**Comment by mattcaswell** [comment]
New fixup pushed addressing another travis issue. Also fixed the nit noted by @slontis.

**Comment by mattcaswell** [comment]
Rebased now that #13317 has gone in.

Ping?

**Comment by t8m** [comment]
You didn't push the rebase.

**Comment by mattcaswell** [comment]
> You didn't push the rebase.

Hmm...the thing is...I did...its just not showing here for some reason.

**Comment by mattcaswell** [comment]
Here's the branch in my personal github repo:

https://github.com/mattcaswell/openssl/commits/no-ssl-low-level-dh

It shows the head commit as being 0fbe6a3. But in this PR based on that branch, the head commit is still showing as c38a675.

WTF?

**Comment by t8m** [comment]
Heh, closed with the intention to reopen but it is no longer possible. @mattcaswell perhaps you could reopen? If not, just open a new PR.

**Comment by mattcaswell** [comment]
Nope, it won't let me reopen it either. Weird.

So, I opened a new PR instead: #13368
